ElectricalCircuitSimulator/SimulationEngine/include/Matrix.h

```diff
@@ -2,72 +2,101 @@
-                m_oMatrix = new double* [0];
+                m_pMatrix = new double* [0];
             };
             Matrix(const int iRows, const int iColumns);
+            Matrix(const Matrix& oMatrix);
+            Matrix(Matrix&&) = delete;
 
             ~Matrix() {
-                delete m_oMatrix;
+                delete m_pMatrix;
             }
 
-            int getNumRows() const { 
+            Matrix& operator=(const Matrix&) = delete;
+            Matrix& operator=(Matrix&&) = delete;
+
+            double& operator()(const int iRow, const int iColumn) const {
+                return m_pMatrix[iRow][iColumn];
+            }
+
+            int getNumRows() const {
                 return m_iNumRows;
             }
             int getNumColumns() const {
                 return m_iNumColumns;
             }
             double getValue(const int iRow, const int iColumn) const;
             void setValue(const int iRow, const int iColumn, const double dValue);
+            void swapRows(const int iRow1, const int iRow2) {
+                swap(m_pMatrix[iRow1], m_pMatrix[iRow2]);
+            }
+            void swapValues(const int iRow1, const int iColumn1, const int iRow2, const int iColumn2) {
+                swap(m_pMatrix[iRow1][iColumn1], m_pMatrix[iRow2][iColumn2]);
+            }
             void clear();
-            Matrix* clone() const;
             void printMatrix() const;
-            PLU_Factorization* runPLU_Factorization() const;
-            static Matrix* linearSystemSolver(const Matrix& oSourceVector, const PLU_Factorization& oPLU_Factorization);
+            static void linearSystemSolver(const Matrix& oSourceVector, const PLU_Factorization& oPLU, Matrix& oVoltageVector);
 
         private:
 
             int m_iNumRows;
             int m_iNumColumns;
-            double** m_oMatrix; // Outer pointer is rows, inner pointer is columns
+            double** m_pMatrix; // Outer pointer is rows, inner pointer is columns
     };
 
     // Represents a matrix factored into P*A*Q = L*U, where P = Row Permutation Matrix, and Q = Column Permutation Matrix
-    class PLU_Factorization
+    class PLU_Factorization final
     {
         public:
 
             PLU_Factorization() {
-                m_pL = new Matrix();
-                m_pU = new Matrix();
-                m_pP = new Matrix();
-                m_pQ = new Matrix();
-            };
+                m_pL = make_unique<Matrix>();
+                m_pU = make_unique<Matrix>();
+                m_pP = make_unique<Matrix>();
+                m_pQ = make_unique<Matrix>();
+            }
+            PLU_Factorization(const Matrix& oConductanceMatrix) {
+                runPLU_Factorization(oConductanceMatrix);
+            }
 
-            ~PLU_Factorization() {
-                delete m_pL;
-                delete m_pU;
-                delete m_pP;
-                delete m_pQ;
-            };
+            Matrix& getL() const {
```

Comment:
Bad! Non-const reference returned from a const function

`const Matrix& getL() const`

same for the others